File: Difficulty: Hard/Alien Dictionary/alien-dictionary.py
from collections import defaultdict, deque
class Solution:
    def findOrder(self, words):
        # code here
        inorder = {}
        for word in words:
            for ch in word:
                inorder[ch] = 0
        adjList = defaultdict(list)
        
        for i in range(1, len(words)):
            word2 = words[i]
            word1 = words[i-1]
            
            if len(word1) > len(word2):
                flag = True
                j = 0
                while j < len(word2):
                    if word1[j] != word2[j]:
                        flag = False
                        break
                    j += 1
                if flag:
                    return ""
        
            length = min(len(word1), len(word2))
            j = 0
            while j < length:
                if word1[j] != word2[j]:
                    if word2[j] not in adjList[word1[j]]:
                        adjList[word1[j]].append(word2[j])
                        inorder[word2[j]] += 1
                    break
                j += 1
        
        # print(adjList, inorder)
        # return "abcd"
        queue = deque([])
        for i in inorder:
            if inorder[i] == 0:
                queue.append(i)
        # queue = deque()
        ans = ""
        while queue:
            node = queue.popleft()                    
            ans += (node)

            for nnode in adjList[node]:
                inorder[nnode] -= 1
                if inorder[nnode] == 0:
                    queue.append(nnode)
        # print(ans)
        if len(ans) != len(inorder):
            return ""
        return ans

File: Difficulty: Hard/Alien Dictionary/alien-dictionary.java
import java.util.*;

class Solution {
    public String findOrder(String[] words) {
        
        HashMap<Character, HashSet<Character>> graph = new HashMap<>();
        HashMap<Character, Integer> indegree = new HashMap<>();
        
        for (String word : words) {
            for (char ch : word.toCharArray()) {
                graph.putIfAbsent(ch, new HashSet<>());
                indegree.putIfAbsent(ch, 0);
            }
        }
        
        for (int i = 0; i < words.length - 1; i++) {
            String w1 = words[i];
            String w2 = words[i + 1];
            
            if (w1.length() > w2.length() && w1.startsWith(w2))
                return "";
                
            int len = Math.min(w1.length(), w2.length());
            
            for (int j = 0; j < len; j++) {
                char c1 = w1.charAt(j);
                char c2 = w2.charAt(j);
                
                if (c1 != c2) {
                    if (!graph.get(c1).contains(c2)) {
                        graph.get(c1).add(c2);
                        indegree.put(c2, indegree.get(c2) + 1);
                    }
                    break;
                }
            }   
        }
        
        Queue<Character> queue = new LinkedList<>();
        
        for (char ch : indegree.keySet()) {
            if (indegree.get(ch) == 0)
                queue.offer(ch);
        }
        
        StringBuilder ans = new StringBuilder();
        
        while (!queue.isEmpty()) {
            char curr = queue.poll();
            ans.append(curr);
        
            for (char next : graph.get(curr)) {
                indegree.put(next, indegree.get(next) - 1);
                if (indegree.get(next) == 0)
                    queue.offer(next);
            }
        }
        
        if (ans.length() != indegree.size())
            return "";
            
        return ans.toString();
        
    }   
}